src/Drivers/hiopbbpy/BODriver.py

```diff
@@ -17,34 +21,109 @@
+saved_yopt = np.load("yopt_20iter_1000run.npy",allow_pickle=True).item()
+
+prob_type_l = ["LpNorm", "Branin"]
+acq_type_l = ["LCB", "EI"]
+
+mean_obj = {}
+max_obj = {}
+min_obj = {}
+y_opt = {}
+
+retval = 0
+for prob_type in prob_type_l:
+   print()
+   if prob_type == "LpNorm":
+      problem = LpNormProblem(nx, xlimits)
+   else:
+      problem = BraninProblem()
+
+   if prob_type not in mean_obj:
+      mean_obj[prob_type] = {}
+      max_obj[prob_type] = {}
+      min_obj[prob_type] = {}
+      y_opt[prob_type] = {}
+
+   for acq_type in acq_type_l:
+      if acq_type not in mean_obj[prob_type]:
+         mean_obj[prob_type][acq_type] = 0
+         max_obj[prob_type][acq_type] = -np.inf
+         min_obj[prob_type][acq_type] = np.inf
+         y_opt[prob_type][acq_type] = np.zeros(num_repeat)
+
+      print("Problem name: ", problem.name)
+      print("Acquisition type: ", acq_type)
+   
+      for n_repeat in range(num_repeat):
+         print("Run: ", n_repeat, "/", num_repeat)
+         ### initial training set
+         x_train = problem.sample(n_samples)
+         y_train = problem.evaluate(x_train)
+
+         ### Define the GP surrogate model
+         gp_model = smtKRG(theta, xlimits, nx)
+         gp_model.train(x_train, y_train)
+   
+         # Instantiate and run Bayesian Optimization
+         bo = BOAlgorithm(gp_model, x_train, y_train, acquisition_type = acq_type) #EI or LCB
+         bo.optimize(problem)
+         
+         # Retrieve optimal objec
+         y_opt[prob_type][acq_type][n_repeat] = bo.getOptimalObjective()
+         
+         mean_obj[prob_type][acq_type] += y_opt[prob_type][acq_type][n_repeat]
+         max_obj[prob_type][acq_type] = max(max_obj[prob_type][acq_type], y_opt[prob_type][acq_type][n_repeat])
+         min_obj[prob_type][acq_type] = min(min_obj[prob_type][acq_type], y_opt[prob_type][acq_type][n_repeat])
+
+#if(num_repeat >= 1000 ):
+#   np.save("yopt_20iter_1000run.npy", y_opt)
+
+# Define percentiles
+left_percentile = 5  # or 1
+right_percentile = 100 - left_percentile  # 95 or 99
+
+print("Summary:")
+for prob_type in prob_type_l:
+   for acq_type in acq_type_l:
+      allowed_error = max(1e-6, 0.01*(saved_max_obj[prob_type][acq_type]-saved_min_obj[prob_type][acq_type]))
+
+      mean_obj[prob_type][acq_type] /= num_repeat
+      print("(Min,Mean,Max) Opt.Obj for", prob_type, "-", acq_type, ":\t(", min_obj[prob_type][acq_type], ",",mean_obj[prob_type][acq_type], ",", max_obj[prob_type][acq_type], ")")
+   
+      ### verify the results with the saved results
+      left_value = np.percentile(saved_yopt[prob_type][acq_type], left_percentile)
+      right_value = np.percentile(saved_yopt[prob_type][acq_type], right_percentile)
+
+      lb = left_value - allowed_error
+      ub = right_value + allowed_error
+
+      is_failed = (y_opt[prob_type][acq_type] < lb) | (y_opt[prob_type][acq_type] > ub)
+      num_fail = np.sum(is_failed)
+
+      if num_fail > 1:
```

Comment:
Is it unlikely to occur that `num_fail > 1`? The probability of failure is 10% having used 5th and 95th percentiles. The probability that the first two runs of ten experiments fail is 1% but the probability that any two of ten experiments fail is 19.4%. This may seem unintuitive but the probability that all succeed is `0.9^10 = 34.9%`. 

Using a binomial distribution calculation I see that their is a `26.4%` likelihood of 2 or more failures of 10 when the likelihood of failure for a single event is 10%.

 I'd suggest we use something like `num_fail > 7` which has a `3.7e-7` probability or `num_fail > 8` which has a `9.1e-9` probability.

Reply:
This is a good point! I also think `num_fail > 1` is too easy to reach, i.e., the probability is 19.4%, and all we can do is to rerun the test till it succeeds. instead of using `num_fail > 7`, I am thinking of using 1% percentile, where any two of ten experiments fail is 0.415%

Reply:
I agree, either 3 or more failures to be within 1st and 99th or 2 or more failures within 0.5 and 99.5 percentiles (probability is 0.00426). I think the first is safer given you used only 1k samples to compute the percentiles.